### PR TITLE
Lean on SSLContext on CPython versions that support it

### LINKED ISSUE #2191: Enable hostname_checks_common_name on Python versions that support it
We'll have to wait on the resolution of https://bugs.python.org/issue43522, and then based on that decide how we detect the feature:

 * use an hypothetical new flag?
 * 3.10?
 * 3.7.x, 3.8.x, 3.9.x and 3.10 since this fix will be backported?

We'll have to wait for https://github.com/python/cpython/pull/24899 to land and to be in the 3.10 we use in GitHub Actions, probably when 3.10.0a7 get released.

### PR DESCRIPTION
**Pull request by pquentin**
Closes #2191

### CODE REVIEW

File: src/urllib3/util/ssl_.py [hunk 1]
```diff
@@ -67,6 +67,20 @@ def _is_ge_openssl_v1_1_1(
     )
     PROTOCOL_SSLv23 = PROTOCOL_TLS
 
+    # https://github.com/urllib3/urllib3/issues/2192#issuecomment-821832963
```
Comment (sethmlarson) [inline src/urllib3/util/ssl_.py:70]: Can we put a brief description of why we're doing this so readers don't have to click into a URL for context?

Something short like "SSLContext.hostname_checks_common_name = False didn't work before CPython 3.8.9, 3.9.3, and 3.10"

**Review by sethmlarson — DISMISSED** [review]
This change LGTM, two small comments. Merge when ready!

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -45,6 +45,11 @@ jobs:
             os: ubuntu-latest
             experimental: false
             nox-session: test_brotlipy
+          # Test CPython with a broken hostname_checks_common_name
+          - python-version: 3.9.2
+            os: ubuntu-latest
+            experimental: false
+            nox-session: test-3.8
```
Comment (sethmlarson) [inline .github/workflows/ci.yml:52]: Should be test-3.9?

Comment (pquentin) [reply .github/workflows/ci.yml]: Nice catch. With the fix, we do run the tests with 3.9.2. Wasn't noticed by coverage as Python 3.7 still covers this for now.

File: src/urllib3/util/ssl_.py (hunk 1 above)
Comment (pquentin) [reply src/urllib3/util/ssl_.py:85]: Thanks, added.

**Comment by pquentin** [comment]
Requesting a new review as I'm now checking the OpenSSL version after all.

File: src/urllib3/util/ssl_.py [hunk 3]
```diff
@@ -36,6 +36,19 @@ def _is_ge_openssl_v1_1_1(
...
+def _is_ge_openssl_v1_1_1l(
+    openssl_version_text: str, openssl_version_number: int
+) -> bool:
+    """Returns True for OpenSSL 1.1.1l+ (>=0x101010cf)
+    LibreSSL reports a version number of 0x20000000 for
+    OpenSSL version number so we need to filter out LibreSSL.
+    """
+    return (
+        not openssl_version_text.startswith("LibreSSL")
+        and openssl_version_number >= 0x101010CF
```
Comment (pquentin) [inline src/urllib3/util/ssl_.py:48]: How do I know this? I currently have 1.1.1l on my laptop.

I considered using OPENSSL_VERSION_INFO which is less obscure and more fitting for Python code (which can compare tuples easily) but was lazy. Please tell me if you'd prefer that.

**Comment by pquentin** [comment]
Requesting a new review as the code changed significantly

**Comment by pquentin** [comment]
I've added OpenSSL but this is now an "and", not an "or". Will rework this.

**Pull request closed without merging**